Empty result using ListObjectsV2Command with Prefix containing special characters (`/`)

Hello, and thanks for the really useful project!

I created an object storage key containing `/` (`hls/toto/filename.mp4`). Using `ListObjectsV2Command` with a `Prefix` only works if we don't specify `/`: 
  * `Prefix: 'hls'` correctly returns `hls/toto/filename.mp4`
  * `Prefix: 'hls/'` or `Prefix: 'hls/toto'` does not return anything

Unfortunately I could not debug it because of a dependency issue when compiling the project (`mvn compile`):

```
[ERROR] Failed to execute goal on project s3ninja: Could not resolve dependencies for project com.scireum:s3ninja:jar:DEVELOPMENT-SNAPSHOT: Failed to collect dependencies at com.scireum:sirius-web:jar:dev-35.0.0 -> com.scireum:sirius-kernel:jar:dev-23.0.0 -> com.palantir.docker.compose:docker-compose-rule-core:jar:1.3.0 -> com.palantir.conjure.java.runtime:conjure-java-jackson-serialization:jar:4.18.1: Failed to read artifact descriptor for com.palantir.conjure.java.runtime:conjure-java-jackson-serialization:jar:4.18.1: Could not transfer artifact com.palantir.conjure.java.runtime:conjure-java-jackson-serialization:pom:4.18.1 from/to bintray-palantir (https://dl.bintray.com/palantir/releases): authorization failed for https://dl.bintray.com/palantir/releases/com/palantir/conjure/java/runtime/conjure-java-jackson-serialization/4.18.1/conjure-java-jackson-serialization-4.18.1.pom, status: 403 Forbidden -> [Help 1]
```

Thanks!

--- a/pom.xml
+++ b/pom.xml
@@ -16,8 +16,8 @@
     <url>http://s3ninja.net</url>
 
     <properties>
-        <sirius.kernel>dev-23.0.0</sirius.kernel>
-        <sirius.web>dev-35.0.0</sirius.web>
+        <sirius.kernel>dev-23.5.0</sirius.kernel>
+        <sirius.web>dev-39.8.0</sirius.web>
     </properties>
 
     <repositories>

--- a/src/main/java/ninja/Aws4HashCalculator.java
+++ b/src/main/java/ninja/Aws4HashCalculator.java
@@ -60,7 +60,7 @@ public class Aws4HashCalculator {
      * Computes the authentication hash as specified by the AWS SDK for verification purposes.
      *
      * @param webContext the current request to fetch parameters from
-     * @param pathPrefix the path prefix to preped to the {@link S3Dispatcher#getEffectiveURI(WebContext) effective URI}
+     * @param pathPrefix the path prefix to preped to the {@link S3Dispatcher#getEffectiveURI(String) effective URI}
      *                   of the request
      * @return the computes hash value
      * @throws Exception when hashing fails

--- a/src/main/java/ninja/AwsLegacyHashCalculator.java
+++ b/src/main/java/ninja/AwsLegacyHashCalculator.java
@@ -63,7 +63,7 @@ public class AwsLegacyHashCalculator {
      * Computes the authentication hash as specified by the AWS SDK for verification purposes.
      *
      * @param webContext the current request to fetch parameters from
-     * @param pathPrefix the path prefix to preped to the {@link S3Dispatcher#getEffectiveURI(WebContext) effective URI}
+     * @param pathPrefix the path prefix to preped to the {@link S3Dispatcher#getEffectiveURI(String) effective URI}
      *                   of the request
      * @return the computes hash value
      * @throws Exception when hashing fails
@@ -95,7 +95,7 @@ public class AwsLegacyHashCalculator {
             stringToSign.append("\n");
         }
 
-        stringToSign.append(pathPrefix).append('/').append(S3Dispatcher.getEffectiveURI(webContext));
+        stringToSign.append(pathPrefix).append('/').append(S3Dispatcher.getEffectiveURI(webContext.getRawRequestedURI()));
 
         char separator = '?';
         for (String parameterName : webContext.getParameterNames().stream().sorted().collect(Collectors.toList())) {

--- a/src/main/java/ninja/ListFileTreeVisitor.java
+++ b/src/main/java/ninja/ListFileTreeVisitor.java
@@ -55,9 +55,9 @@ class ListFileTreeVisitor extends SimpleFileVisitor<Path> {
     @Override
     public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
         File file = path.toFile();
-        String name = file.getName();
+        String name = StoredObject.decodeKey(file.getName());
 
-        if (!file.isFile() || name.startsWith("$")) {
+        if (!file.isFile() || file.getName().startsWith("$")) {
             return FileVisitResult.CONTINUE;
         }
         if (!markerReached) {

--- a/src/main/java/ninja/NinjaController.java
+++ b/src/main/java/ninja/NinjaController.java
@@ -118,7 +118,7 @@ public class NinjaController extends BasicController {
         try {
             buckets = storage.getBuckets();
         } catch (HandledException e) {
-            UserContext.message(Message.error(e.getMessage()));
+            UserContext.message(Message.error().withTextMessage(e.getMessage()));
         }
         webContext.respondWith()
                   .template("/templates/index.html.pasta",
@@ -151,21 +151,21 @@ public class NinjaController extends BasicController {
         // handle /ui/[bucket]?create
         if (webContext.hasParameter("create")) {
             if (bucket.exists()) {
-                UserContext.message(Message.error("Bucket does already exist."));
+                UserContext.message(Message.error().withTextMessage("Bucket does already exist."));
                 webContext.respondWith().redirectTemporarily("/ui/" + bucket.getEncodedName());
                 return;
             }
 
             bucket.create();
 
-            UserContext.message(Message.info("Bucket successfully created."));
+            UserContext.message(Message.info().withTextMessage("Bucket successfully created."));
             webContext.respondWith().redirectTemporarily("/ui/" + bucket.getEncodedName());
             return;
         }
 
         // from this point on, make sure that the bucket exists
         if (!bucket.exists()) {
-            UserContext.message(Message.error("Bucket does not exist."));
+            UserContext.message(Message.error().withTextMessage("Bucket does not exist."));
             webContext.respondWith().redirectTemporarily("/ui");
             return;
         }
@@ -174,7 +174,7 @@ public class NinjaController extends BasicController {
         if (webContext.hasParameter("make-public")) {
             bucket.makePublic();
 
-            UserContext.message(Message.info("ACLs successfully changed"));
+            UserContext.message(Message.info().withTextMessage("ACLs successfully changed"));
             webContext.respondWith().redirectTemporarily("/ui/" + bucket.getEncodedName());
             return;
         }
@@ -183,7 +183,7 @@ public class NinjaController extends BasicController {
         if (webContext.hasParameter("make-private")) {
             bucket.makePrivate();
 
-            UserContext.message(Message.info("ACLs successfully changed"));
+            UserContext.message(Message.info().withTextMessage("ACLs successfully changed"));
             webContext.respondWith().redirectTemporarily("/ui/" + bucket.getEncodedName());
             return;
         }
@@ -192,7 +192,7 @@ public class NinjaController extends BasicController {
         if (webContext.hasParameter("delete")) {
             bucket.delete();
 
-            UserContext.message(Message.info("Bucket successfully deleted."));
+            UserContext.message(Message.info().withTextMessage("Bucket successfully deleted."));
             webContext.respondWith().redirectTemporarily("/ui");
             return;
         }
@@ -274,7 +274,7 @@ public class NinjaController extends BasicController {
     public void object(WebContext webContext, String bucketName, List<String> idParts) {
         Bucket bucket = storage.getBucket(bucketName);
         if (!bucket.exists()) {
-            UserContext.message(Message.error("Bucket does not exist."));
+            UserContext.message(Message.error().withTextMessage("Bucket does not exist."));
             webContext.respondWith().redirectTemporarily("/ui");
             return;
         }
@@ -282,7 +282,7 @@ public class NinjaController extends BasicController {
         String id = String.join("/", idParts);
         StoredObject object = bucket.getObject(id);
         if (!object.exists()) {
-            UserContext.message(Message.error("Object does not exist."));
+            UserContext.message(Message.error().withTextMessage("Object does not exist."));
             webContext.respondWith().redirectTemporarily("/ui/" + bucket.getEncodedName());
             return;
         }
@@ -291,7 +291,7 @@ public class NinjaController extends BasicController {
         if (webContext.hasParameter("delete")) {
             object.delete();
 
-            UserContext.message(Message.info("Object successfully deleted."));
+            UserContext.message(Message.info().withTextMessage("Object successfully deleted."));
             webContext.respondWith().redirectTemporarily("/ui/" + bucket.getEncodedName());
             return;
         }

--- a/src/main/java/ninja/S3Dispatcher.java
+++ b/src/main/java/ninja/S3Dispatcher.java
@@ -258,11 +258,10 @@ public class S3Dispatcher implements WebDispatcher {
      * As we have to support legacy URIs which have an <tt>/s3</tt> prefix, we cut this here, and
      * also the first "/" and only return the effective URI to process.
      *
-     * @param webContext the current request
+     * @param uri the requested URI
      * @return the effective URI to process
      */
-    public static String getEffectiveURI(WebContext webContext) {
-        String uri = webContext.getRequestedURI();
+    public static String getEffectiveURI(String uri) {
         if (uri.startsWith("/s3")) {
             uri = uri.substring(3);
         }
@@ -270,7 +269,7 @@ public class S3Dispatcher implements WebDispatcher {
             uri = uri.substring(1);
         }
 
-        return Strings.urlEncode(uri).replace("+", "%20").replace("%2F", "/");
+        return uri;
     }
 
     /**
@@ -280,7 +279,7 @@ public class S3Dispatcher implements WebDispatcher {
      * @return a structured {@link S3Request}.
      */
     private static S3Request parseRequest(WebContext webContext) {
-        String uri = getEffectiveURI(webContext);
+        String uri = getEffectiveURI(webContext.getRequestedURI());
 
         // we treat the first parameter without value as query string
         Iterator<String> parameterIterator = webContext.getParameterNames().iterator();

--- a/src/test/java/BaseAWSSpec.groovy
+++ b/src/test/java/BaseAWSSpec.groovy
@@ -171,6 +171,43 @@ abstract class BaseAWSSpec extends BaseSpecification {
         summaries.get(1).getKey() == key2
     }
 
+    // reported in https://github.com/scireum/s3ninja/issues/180
+    def "PUT and then LIST with prefix work as expected"() {
+        given:
+        def bucketName = DEFAULT_BUCKET_NAME
+        def key1 = DEFAULT_KEY + "/Eins"
+        def key2 = DEFAULT_KEY + "/Zwei"
+        def key3 = "a/key/with a different/prefix/Drei"
+        def client = getClient()
+        when:
+        if (client.doesBucketExist(bucketName)) {
+            client.deleteBucket(bucketName)
+        }
+        client.createBucket(bucketName)
+        and:
+        client.putObject(
+                bucketName,
+                key1,
+                new ByteArrayInputStream("Eins".getBytes(Charsets.UTF_8)),
+                new ObjectMetadata())
+        client.putObject(
+                bucketName,
+                key2,
+                new ByteArrayInputStream("Zwei".getBytes(Charsets.UTF_8)),
+                new ObjectMetadata())
+        client.putObject(
+                bucketName,
+                key3,
+                new ByteArrayInputStream("Drei".getBytes(Charsets.UTF_8)),
+                new ObjectMetadata())
+        then:
+        def listing = client.listObjects(bucketName, DEFAULT_KEY + '/')
+        def summaries = listing.getObjectSummaries()
+        summaries.size() == 2
+        summaries.get(0).getKey() == key1
+        summaries.get(1).getKey() == key2
+    }
+
     def "PUT and then DELETE work as expected"() {
         given:
         def bucketName = DEFAULT_BUCKET_NAME

